### PR TITLE
Revive extract dump archive test

### PR DESCRIPTION
**Pull request by dkijania**
Reimplemented patch archive test. Previously it was obsolete due to database constraints which prevents deleting blocks, which was a idea to make invalid database. Currently the design of test is as follow:

1. Create two schemas source and target
2. extract blocks from source database with some gaps
3. apply extracted blocks to target database
4. fill block gaps with missing-block-guardian app
5. run replayer on target database to make sure it's correct.

While developing test, I've added tiny feature which auto-detect if environment in which we run test is CI or local development. I introduced mina-automation library which should be a corner stone for developing tests with single node or small test for corner cases test (which scope is to narrow to promote it to integration test).

### CODE REVIEW

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Review by SanabriaRusso — APPROVED** [review]

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: scripts/archive/missing-blocks-guardian.sh [hunk 1]
```diff
@@ -26,6 +26,8 @@ usage() {
     echo ""
     echo "  Optional:"
     echo "    MISSING_BLOCKS_AUDITOR  Path to the missing-blocks-auditor exe (default: mina-missing-blocks-auditor)"
+    echo "    ARCHIVE_BLOCKS          Path to the archive blocks exe (default: mina-archive-blocks)"
+    echo "    BLOCKS_FORMAT           Block fromat type. Possible values: [precomputed/extensional] (default: precomputed)"
```
Comment (georgeee) [inline scripts/archive/missing-blocks-guardian.sh:30]:
```suggestion
    echo "    BLOCKS_FORMAT           Block format type. Possible values: [precomputed/extensional] (default: precomputed)"
```

File: src/app/extract_blocks/extract_blocks.ml [hunk 2]
```diff
@@ -620,8 +621,21 @@ let main ~archive_uri ~start_state_hash_opt ~end_state_hash_opt ~all_blocks () =
             [%log info] "Writing block with $state_hash"
               ~metadata:
                 [ ("state_hash", State_hash.to_yojson block.state_hash) ] ;
+
+            let network_prefix =
+              network
+              |> Option.map ~f:(fun network -> network ^ "-")
```
Comment (georgeee) [inline src/app/extract_blocks/extract_blocks.ml:627]:
```suggestion
              |> Option.value_map ~default:"" ~f:(fun network -> network ^ "-")
```

Comment (georgeee) [reply src/app/extract_blocks/extract_blocks.ml:627]: And remove the next line

File: src/lib/integration_test_lib/util.ml [hunk 3]
```diff
@@ -2,11 +2,26 @@ open Core
 open Async
 module Timeout = Timeout_lib.Core_time
 
-let run_cmd dir prog args =
+let run_cmd dir prog args ?env () =
+  let list =
+    match env with
+    | None ->
```
Comment (georgeee) [inline src/lib/integration_test_lib/util.ml:8]: Suggest the following rewrite for the entire `let list = ...`:

```suggestion
match env with
    | Some (`Extend list) ->List.map list ~f:(fun (key, value) -> `String (key ^ "=" ^ value))
    | _ -> []
```

File: src/lib/precomputed_block/precomputed_block.ml [hunk 4]
```diff
@@ -0,0 +1,296 @@
+(* precomputed_block.ml *)
+
+open Core
+open Async
+
+(* Precomputed_block.t type has changed from mainnet, too hard to recreate old type
```
Comment (georgeee) [inline src/lib/precomputed_block/precomputed_block.ml:6]: Do you mean current `compatible` diverged from actual `mainnet`? Or the comment isn't up-to-date?

File: src/lib/precomputed_block/precomputed_block.ml [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                  Consensus_state.min_window_density consensus_state
+              ; sub_window_densities =
+                  Consensus_state.sub_window_densities consensus_state
+              ; total_currency = Consensus_state.total_currency consensus_state
+              }
+          }
+      }
+  }
+
+(* Couldn't get the deriver to generate a valid of_yojson implementation here for some reason, so we hand=write one instead *)
```
Comment (georgeee) [inline src/lib/precomputed_block/precomputed_block.ml:84]: I wonder, what was the reason?

File: src/lib/precomputed_block/precomputed_block.ml (hunk 4 above)
Comment (georgeee) [reply src/lib/precomputed_block/precomputed_block.ml:6]: It's not clear to me why this new file is in this PR

File: src/test/mina_automation/docker.ml [hunk 6]
```diff
@@ -0,0 +1,130 @@
+open Core
+open Integration_test_lib
```
Comment (georgeee) [inline src/test/mina_automation/docker.ml:2]: Can new files from `mina_automation` be extracted to a seperate PR, and perhaps a README to be added?

It's hard to understand what's happening here, and I'd prefer to review it separately.

**Review by georgeee — CHANGES_REQUESTED** [review]

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/test/archive/patch_archive_test/patch_archive_test.ml [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+  let%bind extensional_files =
+    Sys.ls_dir output_folder
+    >>= Deferred.List.map ~f:(fun e ->
+            Deferred.return (output_folder ^ "/" ^ e) )
+  in
+
+  let n =
+    List.range 0 missing_blocks_count
+    |> List.map ~f:(fun _ -> Random.int (List.length extensional_files))
```
Comment (georgeee) [inline src/test/archive/patch_archive_test/patch_archive_test.ml:80]:
```suggestion
    List.init missing_blocks_count ~f:(fun _ -> Random.int (List.length extensional_files))
```

File: src/test/archive/patch_archive_test/sql.ml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+         {sql| SELECT state_hash from blocks
+       |sql} )
+      ()
+
+  let run (module Conn : Caqti_async.CONNECTION) ~state_hash =
+    Conn.find
+      (Caqti_request.find Caqti_type.string Caqti_type.int
+         {sql| SELECT id from blocks where state_hash = ?
+       |sql} )
+      state_hash
```
Comment (georgeee) [inline src/test/archive/patch_archive_test/sql.ml:14]: Looks like this model isn't used in the PR

**Review by georgeee — APPROVED** [review]
Left a couple of minor comments

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me
